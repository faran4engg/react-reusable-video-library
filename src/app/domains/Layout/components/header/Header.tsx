import { useState, FC, useEffect } from 'react';
import { NavLink } from 'react-router-dom';

import { HeaderProps } from './types';
import Islaamify360Logo from 'assets/img/logo.png';
import { SvgIcon } from 'app/domains/Common/components/svg-icon';

const Header: FC<HeaderProps> = ({ handleIsSidebarOpen }) => {
  // const history = useHistory();
  const twTheme = localStorage.getItem('twTheme');
  const [mode, setMode] = useState(twTheme || 'light');
  const htmlElement = document.querySelector('html')!;

  useEffect(() => {
    if (!localStorage.getItem('twTheme')) {
      htmlElement.classList.add('dark');
      localStorage.setItem('twTheme', 'dark');
      setMode('dark');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const toggleDark = () => {
    if (localStorage.getItem('twTheme') === 'dark') {
      htmlElement.classList.remove('dark');
      localStorage.setItem('twTheme', 'light');
      setMode('light');
    } else {
      htmlElement.classList.add('dark');
      localStorage.setItem('twTheme', 'dark');
      setMode('dark');
    }
  };

  return (
    <header className="z-40 py-2 shadow-md lg:shadow-none shadow-bottom bg-light-mode dark:bg-dark-mode">
      <div className="container flex justify-between items-center px-3 mx-auto h-full text-theme-color-light dark:text-theme-color-dark">
        {/* <!-- Mobile hamburger --> */}
        <button
          onClick={handleIsSidebarOpen}
          className="p-1 -ml-1 rounded-md lg:hidden focus:outline-none focus:shadow-outline-green-200"
          aria-label="Menu"
        >
          <SvgIcon className="w-6 h-6" icon="MenuIcon" />
        </button>

        <div className="flex flex-1 justify-center h-10">
          <NavLink to="/home" className="block lg:hidden">
            <img
              loading="lazy"
              className="h-10 cursor-pointer lg:ml-0"
              src={Islaamify360Logo}
              alt="Islaamify360-Logo"
              width="133"
              height="40"
            />
          </NavLink>
        </div>
        <ul className="flex flex-shrink-0 items-center space-x-6">
          {/* <li className="relative text-theme-color-light dark:text-theme-color-dark">
            <button
              onClick={() => history.push('/search')}
              className="relative align-middle rounded-md focus:outline-none"
              aria-label="Notifications"
              aria-haspopup="true"
            >
              <SvgIcon className="w-5 h-5" icon="SearchIcon" />
            </button>
          </li> */}
          {/* <!-- Theme toggler --> */}
          <li className="relative text-theme-color-light dark:text-theme-color-dark">
            <button
              onClick={toggleDark}
              className="relative align-middle rounded-md focus:outline-none"
              aria-label="Notifications"
              aria-haspopup="true"
            >
              {mode === 'light' ? (
                <span className="mx-auto w-5 h-5 animate-spin">ðŸŒš</span>
              ) : (
                <span className="mx-auto w-5 h-5 animate-spin">ðŸ’¡</span>
              )}

              {/* <!-- Notification badge --> */}
              <span className="inline-block absolute top-0 right-0 w-3 h-3 bg-red-600 rounded-full border-2 border-white animate-ping transform translate-x-1 -translate-y-1 dark:border-gray-800" />
            </button>
          </li>
        </ul>
      </div>
    </header>
  );
};

export default Header;
