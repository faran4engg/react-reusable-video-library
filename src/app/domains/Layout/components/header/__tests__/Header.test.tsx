import React from 'react';
import { act, create } from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';

import Header from '../Header';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const renderer = createRenderer();

describe('<Header />', () => {
  const handleIsSidebarOpenMock = jest.fn();
  it('should render and match the snapshot', () => {
    renderer.render(<Header handleIsSidebarOpen={handleIsSidebarOpenMock} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  describe('toggleDark', () => {
    const renderComponent = () =>
      create(<Header handleIsSidebarOpen={handleIsSidebarOpenMock} />);

    it('should set dark mode when clicked', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          getItem: jest.fn(() => 'light'),
          setItem: jest.fn(() => null),
        },
        writable: true,
      });
      const component = renderComponent();
      const root = component.root;
      const toggleEl = root.findAllByType('button')[1];
      act(() => {
        toggleEl.props.onClick();
      });
      expect(window.localStorage.setItem).toHaveBeenCalled();
      expect(window.localStorage.setItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.setItem).toHaveBeenCalledWith(
        'twTheme',
        'dark',
      );
      expect(component).toMatchSnapshot();
    });

    it('should set light mode when clicked', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          getItem: jest.fn(() => 'dark'),
          setItem: jest.fn(() => null),
        },
        writable: true,
      });

      const component = renderComponent();
      const root = component.root;
      const toggleEl = root.findAllByType('button')[1];
      act(() => {
        toggleEl.props.onClick();
      });
      expect(window.localStorage.setItem).toHaveBeenCalled();
      expect(window.localStorage.setItem).toHaveBeenCalledTimes(1);
      expect(window.localStorage.setItem).toHaveBeenCalledWith(
        'twTheme',
        'light',
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('MenuIcon', () => {
    const renderComponent = () =>
      create(<Header handleIsSidebarOpen={handleIsSidebarOpenMock} />);

    it('should open side bar', () => {
      const component = renderComponent();
      const root = component.root;
      const MenuEl = root.findAllByType('button')[0];
      act(() => {
        MenuEl.props.onClick();
      });

      expect(handleIsSidebarOpenMock).toHaveBeenCalledTimes(1);
      expect(handleIsSidebarOpenMock).toHaveBeenCalledWith(true);
    });
  });

  describe('Logo Icon', () => {
    const renderComponent = () =>
      create(<Header handleIsSidebarOpen={handleIsSidebarOpenMock} />);

    it('should redirect to /', () => {
      const component = renderComponent();
      const root = component.root;
      const MenuEl = root.findAllByType('img')[0];
      act(() => {
        MenuEl.props.onClick();
      });

      expect(mockHistoryPush).toHaveBeenCalledTimes(1);
      expect(mockHistoryPush).toHaveBeenCalledWith('/');
    });
  });
});
