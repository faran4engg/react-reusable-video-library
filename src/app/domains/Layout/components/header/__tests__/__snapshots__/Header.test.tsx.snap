// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render and match the snapshot 1`] = `
<header
  className="z-40 py-2 shadow-md lg:shadow-none shadow-bottom bg-light-mode dark:bg-dark-mode"
>
  <div
    className="container flex justify-between items-center px-3 mx-auto h-full text-theme-color-light dark:text-theme-color-dark"
  >
    <button
      aria-label="Menu"
      className="p-1 -ml-1 rounded-md lg:hidden focus:outline-none focus:shadow-outline-green-200"
      onClick={[MockFunction]}
    >
      <SvgIcon
        className="w-6 h-6"
        icon="MenuIcon"
      />
    </button>
    <div
      className="flex flex-1 justify-center h-10"
    >
      <UNDEFINED
        className="block lg:hidden"
        to="/home"
      >
        <img
          alt="Islaamify360-Logo"
          className="h-10 cursor-pointer lg:ml-0"
          height="40"
          loading="lazy"
          src="logo.png"
          width="133"
        />
      </UNDEFINED>
    </div>
    <ul
      className="flex flex-shrink-0 items-center space-x-6"
    >
      <li
        className="relative text-theme-color-light dark:text-theme-color-dark"
      >
        <button
          aria-haspopup="true"
          aria-label="Notifications"
          className="relative align-middle rounded-md focus:outline-none"
          onClick={[Function]}
        >
          <span
            className="mx-auto w-5 h-5 animate-spin"
          >
            ðŸŒš
          </span>
          <span
            className="inline-block absolute top-0 right-0 w-3 h-3 bg-red-600 rounded-full border-2 border-white animate-ping transform translate-x-1 -translate-y-1 dark:border-gray-800"
          />
        </button>
      </li>
    </ul>
  </div>
</header>
`;

exports[`<Header /> toggleDark should set dark mode when clicked 1`] = `
<header
  className="z-40 py-2 shadow-md lg:p-4 lg:shadow-none shadow-bottom bg-light-mode dark:bg-dark-mode"
>
  <div
    className="container flex justify-between items-center px-6 mx-auto h-full  text-theme-color-light dark:text-theme-color-dark"
  >
    <button
      aria-label="Menu"
      className="p-1 -ml-1 rounded-md lg:hidden focus:outline-none focus:shadow-outline-green-200"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="w-6 h-6"
      >
        menu.svg
      </svg>
    </button>
    <img
      alt="TeachIslam-Logo"
      className="h-6 cursor-pointer"
      onClick={[Function]}
      src="Teaach-Islam.png"
    />
    <ul
      className="flex flex-shrink-0 items-center space-x-6"
    >
      <li
        className="relative text-theme-color-light dark:text-theme-color-dark"
      >
        <button
          aria-haspopup="true"
          aria-label="Notifications"
          className="relative align-middle rounded-md focus:outline-none"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5"
          >
            sun.svg
          </svg>
          <span
            aria-hidden="true"
            className="inline-block absolute top-0 right-0 w-3 h-3 bg-red-600 rounded-full border-2 border-white animate-ping transform translate-x-1 -translate-y-1 dark:border-gray-800"
          />
        </button>
      </li>
    </ul>
  </div>
</header>
`;

exports[`<Header /> toggleDark should set light mode when clicked 1`] = `
<header
  className="z-40 py-2 shadow-md lg:p-4 lg:shadow-none shadow-bottom bg-light-mode dark:bg-dark-mode"
>
  <div
    className="container flex justify-between items-center px-6 mx-auto h-full  text-theme-color-light dark:text-theme-color-dark"
  >
    <button
      aria-label="Menu"
      className="p-1 -ml-1 rounded-md lg:hidden focus:outline-none focus:shadow-outline-green-200"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="w-6 h-6"
      >
        menu.svg
      </svg>
    </button>
    <img
      alt="TeachIslam-Logo"
      className="h-6 cursor-pointer"
      onClick={[Function]}
      src="Teaach-Islam.png"
    />
    <ul
      className="flex flex-shrink-0 items-center space-x-6"
    >
      <li
        className="relative text-theme-color-light dark:text-theme-color-dark"
      >
        <button
          aria-haspopup="true"
          aria-label="Notifications"
          className="relative align-middle rounded-md focus:outline-none"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5"
          >
            moon.svg
          </svg>
          <span
            aria-hidden="true"
            className="inline-block absolute top-0 right-0 w-3 h-3 bg-red-600 rounded-full border-2 border-white animate-ping transform translate-x-1 -translate-y-1 dark:border-gray-800"
          />
        </button>
      </li>
    </ul>
  </div>
</header>
`;
